platooning_tactical_plugin/src/platooning_tactical_plugin.cpp

```diff
@@ -0,0 +1,617 @@
+    tpp.target_time = startTime + relative_time;
+    tpp.x = points[i].x();
+    tpp.y = points[i].y();
+    tpp.yaw = yaws[i];
+
+    tpp.controller_plugin_name = "default";
+    tpp.planner_plugin_name = plugin_discovery_msg_.name;
+
+    traj.push_back(tpp);
+  }
+
+  return traj;
+}
+
+std::vector<PointSpeedPair> PlatooningTacticalPlugin::maneuvers_to_points(const std::vector<cav_msgs::Maneuver>& maneuvers,
+                                                                      double max_starting_downtrack,
+                                                                      const carma_wm::WorldModelConstPtr& wm)
+{
+  std::vector<PointSpeedPair> points_and_target_speeds;
+  std::unordered_set<lanelet::Id> visited_lanelets;
+
+  bool first = true;
+  ROS_DEBUG_STREAM("VehDowntrack: " << max_starting_downtrack);
+  for (const auto& manuever : maneuvers)
+  {
+    if (manuever.type != cav_msgs::Maneuver::LANE_FOLLOWING)
+    {
+      throw std::invalid_argument("Platooning tactical plugin does not support this maneuver type");
+    }
+
+    if (manuever.lane_following_maneuver.parameters.neogition_type == cav_msgs::ManeuverParameters::NO_NEGOTIATION)
+    {
+      throw std::invalid_argument("NO_NEGOTIATION maneuver is not supported by this plugin");
+    }
+
+
+    if (manuever.lane_following_maneuver.parameters.neogition_type == cav_msgs::ManeuverParameters::PLATOONING)
+    {
+      if ( manuever.lane_following_maneuver.start_speed != manuever.lane_following_maneuver.end_speed)
+      {
+        throw std::invalid_argument("Invalid Platooning Maneuver");
+      }
+    }
+
+    cav_msgs::LaneFollowingManeuver lane_following_maneuver = manuever.lane_following_maneuver;
+
+    double starting_downtrack = lane_following_maneuver.start_dist;
+    if (first)
+    {
+      if (starting_downtrack > max_starting_downtrack)
+      {
+        starting_downtrack = max_starting_downtrack;
+      }
+      first = false;
+    }
+
+    ROS_DEBUG_STREAM("Used downtrack: " << starting_downtrack);
+
+    auto lanelets = wm->getLaneletsBetween(starting_downtrack, lane_following_maneuver.end_dist, true);
+
+    ROS_DEBUG_STREAM("Maneuver");
+    std::vector<lanelet::ConstLanelet> lanelets_to_add;
+    for (auto l : lanelets)
+    {
+      ROS_DEBUG_STREAM("Lanelet ID: " << l.id());
+      if (visited_lanelets.find(l.id()) == visited_lanelets.end())
+      {
+        lanelets_to_add.push_back(l);
+        visited_lanelets.insert(l.id());
+      }
+    }
+
+    lanelet::BasicLineString2d route_geometry = carma_wm::geometry::concatenate_lanelets(lanelets_to_add);
+
+    first = true;
+    for (auto p : route_geometry)
+    {
+      if (first && points_and_target_speeds.size() != 0)
+      {
+        first = false;
+        continue;  // Skip the first point if we have already added points from a previous maneuver to avoid duplicates
+      }
+      PointSpeedPair pair;
+      pair.point = p;
+      pair.speed = lane_following_maneuver.end_speed;
```

Comment:
This will be the baseline speed the control plugin modifies to maintain the gap, right? Just making sure I understand the data flow.

Reply:
Yes, that is correct.